fairscale/clib/fused_adam_cuda/fused_adam_cuda_kernel.cu

```diff
@@ -135,24 +134,62 @@ void fused_adam_cuda(
     cudaStream_t stream = at::cuda::getCurrentCUDAStream();
 
     size_t tl_sz = tensor_lists.size();
-    AT_ASSERTM(tl_sz == 4, "expected tensor lists of size 4");
-
-    // check that the model and gradients are FP32
-    AT_ASSERTM(tensor_lists[0][0].scalar_type() == at::ScalarType::Float);
-    AT_ASSERTM(tensor_lists[3][0].scalar_type() == at::ScalarType::Float);
-    multi_tensor_apply<4>(
-        BLOCK_SIZE,
-        chunk_size,
-        noop_flag,
-        tensor_lists,
-        AdamFunctor<4, float, float>(),
-        beta1,
-        beta2,
-        eps,
-        grad_scale,
-        step_size,
-        (adamMode_t) mode,
-        decay
-    );
+    assert(tl_sz == 4 || tl_sz == 5);
+
+    if(tl_sz == 5) {
+        // Mixed precision case
+        assert(tensor_lists[0][0].scalar_type() == at::ScalarType::Float);
+        assert(tensor_lists[3][0].scalar_type() == at::ScalarType::Half);
+        assert(tensor_lists[4][0].scalar_type() == at::ScalarType::Half);
+        multi_tensor_apply<5>(
+            BLOCK_SIZE,
+            chunk_size,
+            noop_flag,
+            tensor_lists,
+            AdamFunctor<5, float, at::Half>(),
+            beta1,
+            beta2,
+            eps,
+            grad_scale,
+            step_size,
+            (adamMode_t) mode,
+            decay
+        );
+    } else {    // tl_sz == 4
```

Comment:
maybe add an explicit comment for the setting here, similar to "mix precision case" for the above :)

Reply:
Great idea! Done!